## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [199]:
import pandas as pd

client_df = pd.read_csv('Resources/client_dataset.csv')

client_df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0   $1,096.80    $762.71        $7.50  105            1  
1      $24.95     $15.09        $1.49   21            0  
2      $13.52      $7.86        $1.68   39            6  
3      $36.42     $24.85        $1.23   29            3  
4     $195.10    $108.17       $46.43   20            1

In [200]:
# View the column names in the data
client_df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [201]:
# Use the describe function to gather some basic statistics
client_df.describe()

client_id      order_id  order_week  order_year  unit_price  unit_cost  \
count $54,639.00    $54,639.00  $54,639.00  $54,639.00  $54,639.00 $54,639.00   
mean  $54,837.87 $5,470,190.03      $11.36   $2,022.99     $136.27     $99.45   
std   $25,487.44 $2,599,806.93       $7.02       $0.08     $183.87    $133.16   
min   $10,033.00 $1,000,886.00       $1.00   $2,022.00       $0.01      $0.01   
25%   $33,593.00 $3,196,372.00       $6.00   $2,023.00      $20.80     $14.84   
50%   $53,305.00 $5,496,966.00      $11.00   $2,023.00      $68.31     $49.89   
75%   $78,498.00 $7,733,869.00      $17.00   $2,023.00     $173.16    $125.57   
max   $99,984.00 $9,998,480.00      $52.00   $2,023.00   $1,396.23    $846.27   

       unit_weight           qty  line_number  
count   $54,639.00    $54,639.00   $54,639.00  
mean         $5.00       $570.26        $2.98  
std          $5.33    $18,795.52        $2.44  
min          $0.00         $0.00        $0.00  
25%          $1.45        $32.00        $1.00  
50%          $3.24        $68.00        $3.00  
75%          $6.89       $170.00        $5.00  
max         $46.43 $3,958,244.00        $9.00

In [202]:
# Use this space to do any additional research
# and familiarize yourself with the data.
client_df.dtypes
client_df.count()
client_df['client_id'].nunique()
client_df["unit_price"].mean()


136.26720694009774

In [203]:
# What three item categories had the most entries?
client_df['category'].value_counts().head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [204]:
# For the category with the most entries,
# which subcategory had the most entries?
client_df.loc[client_df['category'] == 'consumables', :].value_counts('subcategory').head(1)


subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [205]:
# Which five clients had the most entries in the data?

client_df['client_id'].value_counts().head(5)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [206]:
# Store the client ids of those top 5 clients in a list.
top_5_clients = client_df['client_id'].value_counts().head(5).index.tolist()
print(top_5_clients)


[33615, 66037, 46820, 38378, 24741]


In [207]:
# How many total units (the qty column) did the
# client with the most entries order order?
print(client_df.loc[client_df['client_id'] == top_5_clients[0], 'qty'].sum())



64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [208]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty

#created a new column called subtotal
client_df['subtotal'] = client_df['unit_price'] * client_df['qty']

#created a new dataframe with the columns client_id, unit_price, qty, and subtotal
subtotal_df = client_df[['client_id','unit_price' , 'qty', 'subtotal']]

#sorted the subtotal column in descending order
subtotal_sorted_df = subtotal_df.sort_values(by='subtotal', ascending=False)

#converted the subtotal column to an integer
subtotal_sorted_df['subtotal'] = subtotal_sorted_df['subtotal'].astype(int)

#displayed the top 5 rows of the dataframe
subtotal_sorted_df.head()



client_id  unit_price      qty   subtotal
21801      78965     $147.66  3958244  584474309
15112      85931     $323.42   637318  206121387
23929      38473     $674.94   219645  148247196
25972      86721     $100.40   977121   98102948
11740      84344     $211.44   389960   82453142

In [209]:
# Create column that calculates the total weight of order
client_df['total_weight'] = client_df['unit_weight'] * client_df['qty']

#Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.

# Define a function
def shipping_price(total_weight):
    if total_weight > 50:
        return total_weight * 7
    else:
        return total_weight * 10
    


# Apply the function to shipping_price column
client_df['shipping_price'] = client_df['total_weight'].apply(shipping_price)

In [210]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
client_df['total_price'] = (client_df['subtotal'] + client_df['shipping_price']) * 1.0925


client_df[['client_id', 'subtotal', 'shipping_price', 'total_price']].head()



client_id    subtotal  shipping_price  total_price
0      58515 $115,164.00       $5,512.50  $131,839.08
1      37609     $523.95         $312.90      $914.26
2      57113     $527.28         $458.64    $1,077.12
3      46554   $1,056.18         $356.70    $1,543.57
4      92089   $3,902.00       $6,500.20   $11,364.40

In [211]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).

client_df['line_cost'] = client_df['unit_cost'] * client_df['qty'] + client_df['shipping_price']
client_df[['client_id', 'unit_cost', 'qty', 'shipping_price', 'line_cost', 'subtotal']].head()


client_id  unit_cost  qty  shipping_price  line_cost    subtotal
0      58515    $762.71  105       $5,512.50 $85,597.05 $115,164.00
1      37609     $15.09   21         $312.90    $629.79     $523.95
2      57113      $7.86   39         $458.64    $765.18     $527.28
3      46554     $24.85   29         $356.70  $1,077.35   $1,056.18
4      92089    $108.17   20       $6,500.20  $8,663.60   $3,902.00

In [212]:
# Create a column for the profit of
# each line using line cost and line price
client_df['line_profit'] = client_df['subtotal'] - client_df['line_cost']
client_df[['client_id', 'line_cost', 'subtotal', 'line_profit']].head()


client_id  line_cost    subtotal  line_profit
0      58515 $85,597.05 $115,164.00   $29,566.95
1      37609    $629.79     $523.95     $-105.84
2      57113    $765.18     $527.28     $-237.90
3      46554  $1,077.35   $1,056.18      $-21.17
4      92089  $8,663.60   $3,902.00   $-4,761.60

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [213]:
# Check your work using the totals above
specific_order_ids = [2742071, 2173913, 6128929]

filtered_id_df = client_df[client_df['order_id'].isin(specific_order_ids)]

order_id_df = filtered_id_df[['order_id','subtotal', 'shipping_price','total_price']].groupby('order_id').sum()
order_id_df.head()

subtotal  shipping_price  total_price
order_id                                         
2173913  $109,440.76      $39,198.80  $162,388.72
2742071  $127,799.33      $12,074.26  $152,811.90
6128929  $688,564.27     $156,690.87  $923,441.24

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [214]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.

# Set the display format for floating-point numbers
pd.options.display.float_format = '${:,.2f}'.format

# Your code here
client_df[['client_id', 'total_price']].groupby('client_id').sum().sort_values(by='total_price', ascending=False).head(5)



total_price
client_id                
78965     $655,923,670.06
85931     $237,417,001.42
38473     $183,552,571.81
86721     $181,266,297.07
84344     $106,648,111.60

In [215]:
# Create a summary DataFrame showing the totals for the
# top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.

summary_df = client_df[['client_id', 'qty', 'shipping_price', 'subtotal', 'line_profit']].groupby('client_id').sum()
summary_df = summary_df.sort_values(by='line_profit', ascending=False).head(5)
summary_df


qty  shipping_price        subtotal     line_profit
client_id                                                         
78965      4033302   $5,340,756.88 $595,047,041.80 $186,476,214.21
85931       644423  $10,122,549.19 $207,192,783.92  $79,069,911.32
24741       239862   $5,126,448.37  $70,176,885.25  $24,487,985.54
38473       402770   $4,391,003.73 $163,620,503.65  $23,419,658.51
84344       466968   $4,746,583.13  $92,871,825.66  $13,935,414.20

In [217]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.


summary_df['qty'] = summary_df['qty'].astype(int)
summary_df['Shipping Costs'] = summary_df['shipping_price'].astype(int) / 1000000
summary_df['Total'] = summary_df['subtotal'].astype(int) / 1000000
summary_df['Profit'] = summary_df['line_profit'].astype(int) / 1000000
summary_df = summary_df[['qty', 'Shipping Costs', 'Total', 'Profit']]
summary_df.head()


qty  Shipping Costs   Total  Profit
client_id                                         
78965      4033302           $5.34 $595.05 $186.48
85931       644423          $10.12 $207.19  $79.07
24741       239862           $5.13  $70.18  $24.49
38473       402770           $4.39 $163.62  $23.42
84344       466968           $4.75  $92.87  $13.94

In [218]:
# Sort the updated data by "Total Profit" form highest to lowest
summary_df = summary_df.sort_values(by='Profit', ascending=False)
summary_df.head()

qty  Shipping Costs   Total  Profit
client_id                                         
78965      4033302           $5.34 $595.05 $186.48
85931       644423          $10.12 $207.19  $79.07
24741       239862           $5.13  $70.18  $24.49
38473       402770           $4.39 $163.62  $23.42
84344       466968           $4.75  $92.87  $13.94